sat/apiclient/pcs.py

```diff
@@ -0,0 +1,195 @@
+                                           for err_info, xnames in xnames_by_err.items()])
+
+            return f'{self.message}\n{xname_err_summary}'
+
+
+class PCSClient(APIGatewayClient):
+    """Client for the Power Control Service."""
+    base_resource_path = 'power-control/v1/'
+
+    def set_xnames_power_state(self, xnames, power_state, force=False, recursive=False, prereq=False):
+        """Set the power state of the given xnames.
+
+        Args:
+            xnames (list): the xnames (str) to perform the power operation
+                against.
+            power_state (str): the desired power state. Either "on" or "off".
+            force (bool): if True, disable checks and force the power operation.
+            recursive (bool): if True, power on component and its descendants.
+            prereq (bool): if True, power on component and its ancestors.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError: if the given `power_state` is not one of 'on' or 'off'
+            PCSError: if the attempt to power on/off the given xnames with PCS
+                fails. This exception contains more specific information about
+                the failure, which will be included in its __str__.
+        """
+        allowed_states = {'on', 'off', 'soft-off', 'soft-restart', 'hard-restart', 'init', 'force-off'}
+        power_state = power_state.lower()
+        if power_state not in allowed_states:
+            allowed_states_str = ", ".join("\"" + state + "\"" for state in allowed_states)
+            raise ValueError(f'Invalid power state {power_state} given. Must be {allowed_states_str}')
+
+        if force:
+            if power_state in {'off', 'soft-off'}:
+                power_state = 'force-off'
+            elif power_state == 'soft-restart':
+                power_state = 'hard-restart'
+
+        target_xnames = set()
+        if recursive:
+            hsm_client = HSMClient(self.session)
+
+            try:
+                target_xnames |= set(
+                    component['XName']
+                    for xname in xnames
+                    for component in hsm_client.get_node_components(ancestor=xname)
+                )
+            except APIError as err:
+                raise PCSError(f'Could not retrieve descendent components for xnames: {err}') from err
+        if prereq:
+            hsm_client = HSMClient(self.session)
+            try:
+                target_xnames |= set(
+                    component['XName']
+                    for component in hsm_client.get_all_components()
+                    if any(xname.startswith(component['XName']) for xname in xnames)
+                )
+            except APIError as err:
+                raise PCSError(f'Could not query ancestor components for xnames: {err}') from err
+# Check if target_xnames is empty after the previous if conditions
+        if not target_xnames:
+            target_xnames.update(xnames)
```

Comment:
This doesn't actually fix the problem. For example, imagine this function is called as:

```
pcs_client.set_xnames_power_state('x1000c0s0', 'off', recursive=True)
```

In this case, the xnames of the nodes beneath the given slot/ComputeModule will be added to `target_xnames` (e.g. `['x1000c0s0b0n0', 'x1000c0s0b0n1', 'x1000c0s0b1n0', 'x1000c0s0b1n1']`), but the xname of the slot/ComputeModule itself will not be added, so the slot/ComputeModule itself will not be powered off.

Instead it seems that it would be fine to initialize the `target_xnames` set with the list of xnames passed in as the `xnames` argument to this function.

Reply:
initialized the target_xnames set with list of xnames which are being passed